### PR TITLE
Implement newline filter on config file read in

### LINKED ISSUE #1482: gridcoinresearch.conf formating fix to switch from INVESTOR to SOLO mode
When setting up my Gridcoin Wallet, I had to edit the gridcoinresearch.conf configuration file to reflect my non-standard BOINC program and data paths.

The original file looked like a lot of settings just run together on a single line but word wrapping in my editor. So without messing around with the initial settings, I just I added my three new lines at the end and saved it, so gridcoinresearch.conf looked like:

email=<hidden>addnode=node.gridcoin.us addnode=www.grcpool.com addnode=seeds.gridcoin.ifoggz-network.xyz addnode=ec2-3-81-39-58.compute-1.amazonaws.com addnode=addnode-us-central.cycy.me addnode=gridcoin.ddns.net
boincdatadir=D:\\ProgramData\\BOINC\\
boincappdir=D:\\Program Files\\BOINC\\
PrimaryCPID=b1bae93d553caedaf68fe9b70f03e32d

Spent days not being able to advertise a beacon, and being stuck in INVESTOR mode, After many forum posts focused on supplying the correct email and CPID, I eventually noticed my email address had no space between the end of the .com address and the next setting. I took a guess that maybe the parsing routing was not separating out the email address from the next setting properly. I just added a manual newline (CR/LF?) using Notepad on Windows, so the email was now on its own line, and saved the file conf , and restarted the wallet program. The wallet now displayed my correct CPID address, and the ongoing CPID mismatch in the diagnostics window [Verify CPID is valid: Failed (BOINC CPID and gridcoinreseach.conf CPID do not match)] also fixed itself.

With the CPID now displaying instead of INVESTOR, the advertisebeacon command seems to have worked (I am currently still waiting to see if it took effect in the network).

So the working gridcoinresearch.conf file now looks like:

email=<hidden>
addnode=node.gridcoin.us addnode=www.grcpool.com addnode=seeds.gridcoin.ifoggz-network.xyz addnode=ec2-3-81-39-58.compute-1.amazonaws.com addnode=addnode-us-central.cycy.me addnode=gridcoin.ddns.net
boincdatadir=D:\\ProgramData\\BOINC\\
boincappdir=D:\\Program Files\\BOINC\\
PrimaryCPID=b1bae93d553caedaf68fe9b70f03e32d

Is this incorrect storing of the email address in the gridcoinresearch.conf file common? Is it what is causing lots of problems with many people being stuck in INVESTOR mode? I have never see a post that mentioned this formatting issue within gridcoinresearch.conf.

Suggestions:

1) Change the format of the gridcoinresearch.conf file so every setting is on it's own dedicated line. This would avoid the mixed parsing problem where sometimes multiple settings are combined on a single logical line separated by just spaces, while other times lines need newlines separations. I had tried placing the last three settings (boincdatadir / boincappdir / PrimaryCPID) on the same single logical line with only spaces separating them, but the Wallet threw an error trying to read the file on startup.

2) Add a line to the diagnostics screen to display the user's email address. A misspelled or completely missing email address (due to a read error of the conf file) could then be found more quickly.

### PR DESCRIPTION
**Pull request by jamescowens**
This implements a simple boost stream input filter that changes the line endings on the read in config file to the unix standard (LF). I have tested this on the strange double CR mixed files we have seen in the wild lately and this solves the problem elegantly with little fuss.

### CODE REVIEW

**Comment by jamescowens** [comment]
I tested this by changing the CRLF after a sidestake entry in one of my testnet node's config files to a CRCR with a hex editor.

image

When running the current HEAD of the development branch the following output of getmining info is seen...

{
"blocks": 1171810,
"stakeweight": {
"minimum": 0,
"maximum": 0,
"combined": 0,
"valuesum": 0,
"legacy": 0
},
"netstakeweight": 359647731.4024152,
"netstakingGRCvalue": 4495596.64253019,
"staking": false,
"mining-error": "No coins; ",
"time-to-stake_days": 0,
"expectedtime": 0,
"mining-version": 0,
"mining-created": 0,
"mining-accepted": 0,
"mining-kernels-found": 0,
"kernel-diff-best": 0,
"kernel-diff-sum": 0,
"stake-splitting": {
"stake-splitting-enabled": true,
"stake-splitting-params": {
"min-stake-split-value": 800,
"efficiency": 0.98,
"stake-split-UTXO-size-for-target-efficiency": 800
}
},
"side-staking": {
"side-staking-enabled": true,
"side-staking-allocations": [
]
},
"difficulty": {
"current": 0.4599044190404009,
"target": 0.3444080658808196,
"last-search-interval": 0
},
"errors": "",
"pooledtx": 0,
"testnet": true,
"CPID": "INVESTOR",
"MiningInfo 1": "Staking Only - No Eligible Research Projects",
"MiningInfo 2": "Poll: No current polls",
"MiningInfo 5": "",
"MiningInfo 6": "",
"MiningInfo 7": "",
"MiningInfo 8": ""
}

When this commit is run the following output is seen:

{
"blocks": 1171814,
"stakeweight": {
"minimum": 0,
"maximum": 0,
"combined": 0,
"valuesum": 0,
"legacy": 0
},
"netstakeweight": 364193167.8277673,
"netstakingGRCvalue": 4552414.597847091,
"staking": false,
"mining-error": "No coins; ",
"time-to-stake_days": 0,
"expectedtime": 0,
"mining-version": 0,
"mining-created": 0,
"mining-accepted": 0,
"mining-kernels-found": 0,
"kernel-diff-best": 0,
"kernel-diff-sum": 0,
"stake-splitting": {
"stake-splitting-enabled": true,
"stake-splitting-params": {
"min-stake-split-value": 800,
"efficiency": 0.98,
"stake-split-UTXO-size-for-target-efficiency": 800
}
},
"side-staking": {
"side-staking-enabled": true,
"side-staking-allocations": [
{
"address": "n34ysBaxXxNjNQnDA94eoqfSkaCsZ54Ytk",
"allocation-pct": 25
},
{
"address": "mhit3mxqGZnhPVZQkNZR6LKAS8SmZeuN4D",
"allocation-pct": 10
}
]
},
"difficulty": {
"current": 0.5210246380613924,
"target": 0.6368123912895609,
"last-search-interval": 0
},
"errors": "",
"pooledtx": 0,
"testnet": true,
"CPID": "INVESTOR",
"MiningInfo 1": "Staking Only - No Eligible Research Projects",
"MiningInfo 2": "Poll: No current polls",
"MiningInfo 5": "",
"MiningInfo 6": "",
"MiningInfo 7": "",
"MiningInfo 8": ""
}

Notice how the allocation entries are absent with the HEAD of the dev branch, indicating confusion generated by the double CR, but properly present with this commit.

This should solve the email= entries that still result in INVESTOR status that have dogged some people, as well as org= entries on testnet.

File: src/util.h [hunk 1]
```diff
@@ -28,6 +28,8 @@
 #include <boost/date_time/posix_time/posix_time_types.hpp>
 #include <boost/thread.hpp>
 #include <boost/thread/condition_variable.hpp>
+#include <boost/iostreams/filtering_stream.hpp>
+#include <boost/iostreams/filter/newline.hpp>
```
Comment (cyrossignol) [inline src/util.h]: Can we put these into util.cpp (so they're not included into every file that needs util.h)?

Comment (jamescowens) [reply src/util.h]: Sure.

Comment (jamescowens) [reply src/util.h]: We actually need to break util apart like bitcoin. But that is for another day...

**Review by cyrossignol — APPROVED** [review]
Nice, simple solution 🙂

File: src/util.h (hunk 1 above)
Comment (cyrossignol) [reply src/util.h]: Yep. One day...

**Comment by jamescowens** [comment]
@barton2526 This one should interest you. That pesky org= problem with d7cd was due to this.

**Pull request merged**